backend/src/email/middlewares/email.middleware.ts

```diff
@@ -79,22 +81,157 @@ const previewFirstMessage = async (
 
     if (!message) return res.json({})
 
-    const { body, subject, replyTo } = message
+    const { body, subject, replyTo, from } = message
     return res.json({
       preview: {
         body,
         subject,
         reply_to: replyTo,
+        from,
       },
     })
   } catch (err) {
     return next(err)
   }
 }
 
+/**
+ * Checks that the from address is either the user's email or the default app email
+ */
+const isFromAddressAccepted = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<Response | void> => {
+  const { from } = req.body
+  const userEmail = req.session?.user?.email
+  const defaultEmail = config.get('mailFrom')
+
+  // Since from addresses with display name are accepted, we need to extract just the email address
+  const { name, fromAddress } = parseFromAddress(from)
+
+  if (fromAddress !== userEmail && from !== defaultEmail) {
+    return res.status(400).json({ message: "Invalid 'from' email address." })
+  }
+  res.locals.fromName = name
+  res.locals.from = fromAddress
+
+  return next()
+}
+
+/**
+ * Verifies that the 'from' address provided is valid
+ */
+const existsFromAddress = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<Response | void> => {
+  const { from } = req.body
+  const defaultEmail = config.get('mailFrom')
+  if (from === defaultEmail) return next()
+
+  try {
+    const { from: fromAddress } = res.locals
+    const result = await CustomDomainService.getCustomFromAddress(fromAddress)
+    if (!result) throw new Error('From Address has not been verified.')
+  } catch (err) {
+    return res.status(400).json({ message: err.message })
+  }
+  return next()
+}
+
+/**
+ * Verifies the user's email address to see if it can be used as custom 'from' address
+ */
+const verifyFromAddress = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<Response | void> => {
+  const { from } = req.body
+  const defaultEmail = config.get('mailFrom')
+  if (from === defaultEmail) return next()
+
+  try {
+    const { from: fromAddress } = res.locals
+    await CustomDomainService.verifyFromAddress(fromAddress)
+  } catch (err) {
+    return res.status(400).json({ message: err.message })
+  }
+  return next()
+}
+
+/**
+ * Stores the verified email address that we can use to send out emails.
+ */
+const storeFromAddress = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<Response | void> => {
+  const { from } = req.body
+  const defaultEmail = config.get('mailFrom')
+  if (from === defaultEmail) return res.sendStatus(200)
```

Comment:
Is there a better way of handling the check against defaultEmail?  Worried about users adding donotreply@mail.postman.gov.sg